mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -345,9 +347,179 @@ propagateUnmergeVectorization(T &&dimAndLength,
+      // Protect against broadcasts
+      if (unmergeParams[j] == 1) {
+        continue;
+      }
+
+      auto it = dimToMerge.find(unmergeDims[j]);
+      if (it == dimToMerge.end() || it->second != mergePair ||
+          paramMatcher[unmergeDims[j]] != unmergeParams[j]) {
+        break;
+      }
+      outputDims.push_back(unmergeDims[j]);
+    }
+
+    // Now we know outputDims contain a group of contiguous dimensions
+    i += std::max(size_t(1), outputDims.size());
+
+    // Update the result with the current group for the mergePair key
+    if (outputDims.size() > 1) {
+      std::sort(outputDims.begin(), outputDims.end(),
+                [&](int a, int b) { return reshuffler[a] < reshuffler[b]; });
+      for (auto d : outputDims) {
+        dimToMerge.erase(d);
+      }
+      result[mergePair].push_back(outputDims);
+    }
+  }
+}
+
+/**
+ * Utility function to filter dimensions that have been spuriously added
+ */
+SmallVector<uint32_t>
+filterDummyDimensions(ArrayRef<uint32_t> dims,
+                      DenseSet<uint32_t> &dummyDimensions) {
+  uint32_t inc = 0;
+  SmallVector<uint32_t> newDims(dims);
+  for (auto &d : newDims) {
+    if (dummyDimensions.contains(d)) {
+      inc++;
+    }
+    d += inc;
+  }
+  return newDims;
+}
+
+/** This is an analysis pass to group the lower dimensions of a Merge
+ * transformation into contiguous groups. E.g., if we have a Merge{8,8,3} [0] ->
+ * [0,2,3] and we know that [2,3] are contiguous in the final representation, we
+ * can split the merge group in
+ * [[0] [2,3]]. This information will be used by the vectorizer. E.g., the
+ * vectorizer can fulfill a vectorization by 4, since 8*3=24 is a multiple of 4.
+ * In other words, every group of dimensions is treated as a single group
+ */
+ContiguousMergesMap findContiguousMerges(ArrayAttr transforms,
+                                         ArrayRef<int64_t> outputShape) {
+  DimToMergeMap dimToMerge;
+  ContiguousMergesMap result;
+  DenseSet<uint32_t> dummyDimensions;
+
+  for (TransformMapAttr transformMap :
+       transforms.getAsRange<TransformMapAttr>()) {
+    for (TransformAttr transform : transformMap.getOps()) {
+      auto transformType = transform.getType();
+      ArrayRef<uint32_t> lowerDims = transform.getLowerDims();
+      ArrayRef<uint32_t> upperDims = transform.getUpperDims();
+      ArrayRef<int64_t> params = transform.getParams();
+
+      switch (transformType) {
+      case TransformType::Merge:
+        for (auto d : lowerDims) {
+          dimToMerge[d] = std::make_pair(transformMap, transform);
+        }
+        break;
+      case TransformType::AddDim:
+        dummyDimensions.insert(upperDims[0]);
+        break;
+      case TransformType::Embed: {
+        SmallVector<int64_t> unmergeParams;
+        for (size_t i = 0; i < params.size() - 1; i++) {
+          unmergeParams.push_back(params[i + 1] / params[i]);
+        }
+        unmergeParams.push_back(params.back());
```

Comment:
Missing a division in any case. But also, very annoyingly, that final coefficient doesn't tell you the length of the whole thing. You'll want to look into the sizes on the transform map, I think.

Reply:
Yes, you are right. But I don't think the transform map would tell me either. For instance:
```
#transform_unmerge_2 = #rock.transform_map<affine_map<(d0, d1, d2, d3) -> (d3 + 2*d2 + 4*d1 + 128*d0)>
  by [<Unmerge{2, 2, 32, 128} ["Wo", "Ho", "K", "N"] at [3, 2, 1, 0] -> ["x"] at [0]>]
  bounds = [2, 2, 32, 128] -> [16384]>
```
The map is simply applying the coefficients to the dimensions (I think). Am I wrong? Once I make sure that this is a merge-like transformation, I could either:
- add a `-1` in the `unmergeParams` and make the unmerge logic ignore the -1
- use the bounds to correctly detect the last parameter
Please, let me know if I am missing something.

Reply:
Also, independently by the last parameters I found that although I am grouping the dimensions in the right way, the vectorization of `Embed` still fails to vectorize properly. It seems to me that `Unfold+Embed` struggles to vectorize properly.

Reply:
I'm suggesting using the bounds to work out the last parameter. It's not great, and we probably should move more Embed{} calls to Unmerge{}, but it should do.

Reply:
Perfect, let's try this way. Thanks @krzysz00 !